sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -170,6 +170,8 @@ case class InMemoryTableScanExec(
   override def outputPartitioning: Partitioning = {
     relation.cachedPlan.outputPartitioning match {
       case h: HashPartitioning => updateAttribute(h).asInstanceOf[HashPartitioning]
+      case r: RangePartitioning =>
+        r.copy(ordering = r.ordering.map(updateAttribute(_).asInstanceOf[SortOrder]))
```

Comment:
why not just `updateAttribute(r)`?

Moreover, in order to avoid the same issue in the future with other cases, have you considered doing something like:
```
updateAttribute(relation.cachedPlan.outputPartitioning)
``
?

Reply:
Not all `Partitioning` are `Expression`. Only `HashPartitioning` and `RangePartitioning` are.

Reply:
Good suggestion, thanks @mgaido91.

@viirya Do we need consider below:
`PartitioningCollection` in `InMemoryTableScanExec.outputPartitioning`, which is also `Expression`?
`PartitioningCollection` and `BroadcastPartitioning` in `ReusedExchangeExec.outputPartitioning`?

Reply:
`BroadcastPartitioning`'s `BroadcastMode` contains `Expression`?

Reply:
Oh, like `HashedRelationBroadcastMode`.

Reply:
Hmm, `HashPartitioning` and `RangePartitioning` can affect later sorting and shuffle. But for `BroadcastPartitioning`, seems to me no such benefit.

Reply:
`PartitioningCollection` should be considered. Like below case:
```
spark.conf.set("spark.sql.autoBroadcastJoinThreshold", -1)
spark.conf.set("spark.sql.codegen.wholeStage", false)
val df1 = Seq(1 -> "a", 3 -> "c", 2 -> "b").toDF("i", "j").as("t1")
val df2 = Seq(1 -> "a", 3 -> "c", 2 -> "b").toDF("m", "n").as("t2")
val d = df1.join(df2, $"t1.i" === $"t2.m")
d.cache
val d1 = d.as("t3")
val d2 = d.as("t4")
d1.join(d2, $"t3.i" === $"t4.i").explain
```
```
SortMergeJoin [i#5], [i#54], Inner
:- InMemoryTableScan [i#5, j#6, m#15, n#16]
:     +- InMemoryRelation [i#5, j#6, m#15, n#16], CachedRDDBuilder
:           +- SortMergeJoin [i#5], [m#15], Inner
:              :- Sort [i#5 ASC NULLS FIRST], false, 0
:              :  +- Exchange hashpartitioning(i#5, 10)
:              :     +- LocalTableScan [i#5, j#6]
:              +- Sort [m#15 ASC NULLS FIRST], false, 0
:                 +- Exchange hashpartitioning(m#15, 10)
:                    +- LocalTableScan [m#15, n#16]
+- Sort [i#54 ASC NULLS FIRST], false, 0
   +- Exchange hashpartitioning(i#54, 10)
      +- InMemoryTableScan [i#54, j#55, m#58, n#59]
            +- InMemoryRelation [i#54, j#55, m#58, n#59], CachedRDDBuilder
                  +- SortMergeJoin [i#5], [m#15], Inner
                     :- Sort [i#5 ASC NULLS FIRST], false, 0
                     :  +- Exchange hashpartitioning(i#5, 10)
                     :     +- LocalTableScan [i#5, j#6]
                     +- Sort [m#15 ASC NULLS FIRST], false, 0
                        +- Exchange hashpartitioning(m#15, 10)
                           +- LocalTableScan [m#15, n#16]
```
`Exchange hashpartitioning(i#54, 10)` is extra shuffle.

How do you think?

Reply:
For `PartitioningCollection`, I think it is harder to treat it like `HashPartitioning` and `RangePartitioning` when replacing attributes.

In above example, `PartitioningCollection` contains `HashPartitioning(i#5)` and `HashPartitioning(m#15)`, the output of `InMemoryRelation` is `[i#54, j#55, m#58, n#59]`.  Can we still replace attributes based on the location of attribute in output?

Reply:
@viirya From `updateAttribute`, `relation.cachedPlan.output` and `relation.output` one to one.
``` 
 private def updateAttribute(expr: Expression): Expression = {
    ....
    val attrMap = AttributeMap(relation.cachedPlan.output.zip(relation.output))
    ....
  }
```
It means "[i#54, j#55, m#58, n#59]" corresponds to "[i#5, j#6, m#15, n#16]", so we can always replace `HashPartitioning(i#5)` to `HashPartitioning(i#54)`.
Any idea?

Reply:
Looks correct.